pkg/ha/ha_service.go

```diff
@@ -28,26 +27,64 @@ func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
 	if err != nil {
 		return nil, fmt.Errorf("could not create HA state: %#v", err)
 	}
-	return &Service{
-		state: &State{
-			_mu: sync.RWMutex{},
-		},
-		lockClient:   lockClient,
-		leaseTimeout: timeout,
-		leaseRefresh: refresh,
+
+	service := &Service{
+		state:             make(map[string]*State),
+		lockClient:        lockClient,
+		leaseTimeout:      timeout,
+		leaseRefresh:      refresh,
 		_leaderChangeLock: semaphore.NewWeighted(1),
-	}, nil
+	}
+	go service.haStateSyncer()
+	return service, nil
+}
+
+func (h *Service) haStateSyncer() {
+	for {
+		for clusterName, state := range h.state {
+			// To validate cluster state we use maxTimeInstance instead if state.leader
+			// because if prometheus leader crashes promscale needs to sync leader
+			// using maxTimeSeenInstance and validate if leaderHasChanged condition
+			err := h.validateClusterState(state.minTimeSeen, state.maxTimeSeen, clusterName, state.maxTimeInstance)
+			if err != nil {
+				log.Error("failed to validate cluster %s state %v", clusterName, err)
+			}
+		}
+		time.Sleep(15 * time.Second)
+	}
 }
 
 // checkInsert verifies the samples are from prom leader & in expected time range.
 // 	returns true if the insert is allowed
 func (h *Service) checkInsert(minT, maxT time.Time, clusterName, replicaName string) (bool, error) {
 	// TODO use proper context
-	lockState, err := h.lockClient.checkInsert(context.Background(), clusterName, replicaName, minT, maxT)
+	if h.state[clusterName] == nil {
+		h.state[clusterName] = &State{}
```

Comment:
wasn't there a mutex in State that needs to be initialized? (https://github.com/timescale/promscale/pull/428/files#diff-0756b0d41936dc75d15427359423932e2860ff97542c11aeabd209798031095cL33)